rust/src/aggregator/py_aggregator.rs

```diff
@@ -185,56 +191,62 @@ pub fn spawn_py_aggregator(
 }
 
 pub struct PyAggregatorHandle {
-    pub aggregate_requests: RequestTx<(), Weights>,
-    pub add_weights_requests: RequestTx<Weights, ()>,
+    pub aggregate_requests: RequestTx<(), Result<Bytes, PyAggregatorError<AggregationFailed>>>,
+    pub add_weights_requests: RequestTx<Bytes, Result<(), PyAggregatorError<InvalidWeights>>>,
 }
 
 impl Aggregator for PyAggregatorHandle {
-    type Error = ();
-    type AggregateFut = Pin<Box<dyn Future<Output = Result<Bytes, ()>> + Send>>;
-    type AddWeightsFut = Pin<Box<dyn Future<Output = Result<(), ()>> + Send>>;
+    type Error = PyAggregatorHandleError;
+    type AggregateFut = Pin<Box<dyn Future<Output = Result<Bytes, Self::Error>> + Send>>;
+    type AddWeightsFut = Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send>>;
 
     fn add_weights(&mut self, weights: Bytes) -> Self::AddWeightsFut {
-        let (tx, rx) = oneshot::channel::<()>();
+        let (tx, rx) = oneshot::channel::<Result<(), PyAggregatorError<InvalidWeights>>>();
         let add_weights_requests = self.add_weights_requests.clone();
         Box::pin(async move {
-            add_weights_requests.send((weights, tx)).map_err(|_| ())?;
-            rx.await.map_err(|_| ())
+            add_weights_requests
+                .send((weights, tx))
+                .map_err(|_| ChannelError::Request)?;
+            rx.await
+                .map_err(|_| ChannelError::Response)?
+                .map_err(From::from)
```

Comment:
This method returns a `Self::AddWeightsFut`, ie a `Future<Output = Result<(), PyAggregatorHandleError>`. So this async block should return a `Result<(), PyAggregatorHandleError>`.

`rx.await` returns a `Result<Result<(), PyAggregatorError<InvalidWeights>>, RecvError>` (see [`tokio::sync::oneshot::error::RecvError`](https://docs.rs/tokio/0.2.16/tokio/sync/oneshot/error/struct.RecvError.html)).

So we need to convert:

```rust
Result<Result<(), PyAggregatorError<InvalidWeights>>, RecvError>
```
into
```rust
Result<(), PyAggregatorHandleError>
```

The first `.map_err(|_| ChannelError::Response)` maps

```rust
Result<(), PyAggregatorError<InvalidWeights>>, RecvError>
```
to
```rust
Result<(), PyAggregatorError<InvalidWeights>>, ChannelError>
```

Then comes the `?`, which returns early if the value is an error. There's some magic at play here. Remember that we want to return a `PyAggregatorHandleError`, *not* `ChannelError`, so `?` cannot just return the value. However `PyAggregatorHandleError` is defined as:

```rust
#[derive(Error, Debug)]
pub enum PyAggregatorHandleError {
    #[error("failed to send the request or receive the response")]
    Handle(#[from] ChannelError), // Note the `#[from]` attribute here

    #[error("error while executing Python code")]
    Python(#[from] PythonError),

    #[error("request failed: {0}")]
    Request(String),
}
```

Thanks to the `#[from]` attribute, [`thiserror`](https://github.com/dtolnay/thiserror) generates the following impl:

```rust
impl From<ChannelError> for PyAggregatorHandleError {
    fn from(e: ChannelError) -> Self {
        Self::Handle(e)
    }
}
```
So before returning, `?` is able to convert the `ChannelError` into a `PyAggregatorHandleError`.

To sump up: `.map_err(|_| ChannelError::Response)?` maps

```rust
Result<Result<(), PyAggregatorError<InvalidWeights>>, RecvError>
```
to
```rust
Result<Result<(), PyAggregatorError<InvalidWeights>>, PyAggregatorHandleError>>
```
and:

- if the value is an error, it returns an `Err(PyAggregatorHandleError)`
- if the value is ok, it unwraps it, so we end up with a `Result<(), PyAggregatorError<InvalidWeights>>`


Now we just have to convert this to `Result<(), PyAggregatorHandleError>`. The conversion is easy, because we have:

```rust
impl<E> From<PyAggregatorError<E>> for PyAggregatorHandleError
where
    E: Error,
{
    fn from(e: PyAggregatorError<E>) -> Self {
        match e {
            PyAggregatorError::Python(inner) => Self::Python(inner),
            PyAggregatorError::Request(inner) => Self::Request(format!("{}", inner)),
        }
    }
}
```

So all our `map_err` has to do is call the `From` impl. In fact, the `map_err(From::from)` could be written:


```rust
map_err(|e: PyAggregatorError<InvalidWeights>| From::from(e))`
```
Or even:
```rust
map_err(|e: PyAggregatorError<InvalidWeights>| PyAggregatorHandleError::from(e))`
```

But the closure is redundant: `map_err` just takes a function so these can be shortened:
```rust
map_err(From::from)
// or
map_err(PyAggregatorHandleError::from)
```

Last thing: we can call `From::from` directly here because the compiler is smart enough to know which `From<PyAggregatorError>` impl to call by using the return type of the method. It's a bit less explicit than using `PyAggregatorHandleError::from`, but it's pretty common.

Reply:
Here is the same snipped, with type annotations. Feel free to play with it, changing some types.

```rust
let res_res: Result<Result<(), PyAggregatorError<InvalidWeights>>, oneshot::error::RecvError> = rx.await;

let res_res_mapped: Result<Result<(), PyAggregatorError<InvalidWeights>>, ChannelError,> = res_res.map_err(|_| ChannelError::Response);
let res = res_res_mapped?;
let res_mapped = res.map_err(|e: PyAggregatorError<InvalidWeights>| PyAggregatorHandleError::from(e));

res_mapped
```

Reply:
Wow thanks again for the explanation👍 Now it makes sense to me to just pass the function instead of building a closure that calls the the function inside.